config/templates/stack-template.json

```diff
@@ -1020,139 +1079,305 @@
+        "SubnetId": {
+          "Ref": "{{$controllerSubnetLogicalName}}"
+        }
+      },
+      "Type": "AWS::EC2::SubnetRouteTableAssociation"
+    }
+    {{end}}
+    {{end}}
+    {{end}}
+
+    {{if $.Worker.TopologyPrivate}}
+    {{range $index, $subnet := .Worker.PrivateSubnets}}
+    {{with $workerSubnetLogicalName := printf "Worker%s" $subnet.LogicalName}}
+    ,
+    "{{$workerSubnetLogicalName}}": {
       "Properties": {
+        "AvailabilityZone": "{{$subnet.AvailabilityZone}}",
+        "CidrBlock": "{{$subnet.InstanceCIDR}}",
+        "MapPublicIpOnLaunch": {{not $.Worker.TopologyPrivate}},
         "Tags": [
           {
             "Key": "Name",
-            "Value": "{{$.ClusterName}}-route-table"
+            "Value": "{{$.ClusterName}}-{{$workerSubnetLogicalName}}"
           },
           {
             "Key": "KubernetesCluster",
-            "Value": "{{.ClusterName}}"
+            "Value": "{{$.ClusterName}}"
           }
         ],
-        "VpcId": {{.VPCRef}}
+        "VpcId": {{$.VPCRef}}
       },
-      "Type": "AWS::EC2::RouteTable"
-    },
-    "RouteToInternet": {
+      "Type": "AWS::EC2::Subnet"
+    }
+    ,
+    "{{$workerSubnetLogicalName}}RouteTableAssociation": {
       "Properties": {
-        "DestinationCidrBlock": "0.0.0.0/0",
-        "GatewayId": {
-          "Ref": "InternetGateway"
-        },
-        "RouteTableId": { "Ref" : "RouteTable" }
+        "RouteTableId":
+          {{if not $subnet.RouteTableID}}
+          {"Ref" : "PrivateRouteTable{{$subnet.AvailabilityZoneLogicalName}}"},
+          {{else}}
+          "{{$subnet.RouteTableID}}",
+          {{end}}
+        "SubnetId": {
+          "Ref": "{{$workerSubnetLogicalName}}"
+        }
       },
-      "Type": "AWS::EC2::Route"
-    },
-    "InternetGateway": {
+      "Type": "AWS::EC2::SubnetRouteTableAssociation"
+    }
+    {{end}}
+    {{end}}
+    {{end}}
+    ,
+    "PublicRouteTable": {
       "Properties": {
         "Tags": [
           {
             "Key": "Name",
-            "Value": "{{$.ClusterName}}-igw"
+            "Value": "{{$.ClusterName}}-PublicRouteTable"
           },
           {
             "Key": "KubernetesCluster",
-            "Value": "{{.ClusterName}}"
+            "Value": "{{$.ClusterName}}"
           }
-        ]
+        ],
+        "VpcId": {{$.VPCRef}}
       },
-      "Type": "AWS::EC2::InternetGateway"
+      "Type": "AWS::EC2::RouteTable"
     },
-    "VPCGatewayAttachment": {
+    "PublicRouteToInternet": {
       "Properties": {
-        "InternetGatewayId": {
-          "Ref": "InternetGateway"
-        },
-        "VpcId": {{.VPCRef}}
+        "DestinationCidrBlock": "0.0.0.0/0",
+        "GatewayId": {{$.InternetGatewayRef}},
+        "RouteTableId": {"Ref" : "PublicRouteTable"}
       },
-      "Type": "AWS::EC2::VPCGatewayAttachment"
+      "Type": "AWS::EC2::Route"
     }
+
+    {{if (or .WorkerTopologyPrivate .Etcd.TopologyPrivate .Controller.TopologyPrivate)}}
```

Comment:
Should also add `.Worker.TopologyPrivate` for the case where workers are private yet somehow etcd & controller are public and not using node pools.